numpy/_core/src/npysort/binsearch.cpp

```diff
@@ -64,43 +64,101 @@ binsearch(const char *arr, const char *key, char *ret, npy_intp arr_len,
 {
     using T = typename Tag::type;
     auto cmp = side_to_cmp<Tag, side>::value;
-    npy_intp min_idx = 0;
-    npy_intp max_idx = arr_len;
-    T last_key_val;
 
-    if (key_len == 0) {
+    // Let's handle this corner case first: if length is 0 we return all 0s
+    if (arr_len <= 0) {
+        for (npy_intp i = 0; i < key_len; ++i) {
+            *(npy_intp *)(ret + i * ret_str) = 0;
+        }
         return;
     }
-    last_key_val = *(const T *)key;
 
-    for (; key_len > 0; key_len--, key += key_str, ret += ret_str) {
-        const T key_val = *(const T *)key;
-        /*
-         * Updating only one of the indices based on the previous key
-         * gives the search a big boost when keys are sorted, but slightly
-         * slows down things for purely random ones.
-         */
-        if (cmp(last_key_val, key_val)) {
-            max_idx = arr_len;
-        }
-        else {
-            min_idx = 0;
-            max_idx = (max_idx < arr_len) ? (max_idx + 1) : arr_len;
-        }
+    /*
+    In this binary search, the candidates are in the range [base, base+length]
```

Comment:
```suggestion
    In this binary search, the candidate insertion indices for the jth key are in the range [base_j, base_j+length]
```
(or something similar to make clear each key has its own interval)